pecan/lang/pecan_ast/arith.py

```diff
@@ -4,44 +4,142 @@
 import spot
 
 from pecan.lang.pecan_ast.prog import *
+from pecan.tools.automaton_tools import Substitution, AutomatonTransformer
+from pecan.lang.pecan_ast.bool import *
 
+#TODO: preassign label to expression
 class Add(BinaryExpression):
     def __init__(self, a, b):
         super().__init__(a, b)
+        self.a = a
+        self.b = b
 
     def __repr__(self):
         return '({} + {})'.format(self.a, self.b)
 
+    def evaluate(self, prog):
+        (aut_a, val_a) = self.a.evaluate(prog)
+        (aut_b, val_b) = self.b.evaluate(prog)
+        aut_add = '' #TODO: get automaton
+        # Assuming label "a+b=c"
+        def build_add_formula(self, formula): 
+            return Substitution({'a': val_a, 'b': val_b,'c': '{}{}'.format(self.a,self.b)}).substitute(formula)
+        aut_add = AutomatonTransformer(aut_add, build_add_formula).transform()
+        #TODO: drop val_a, val_b in return
+        return (spot.product(aut_a, spot.product(aut_b, aut_add)), '{}_add_{}'.format(self.a,self.b))
```

Comment:
For here and probably all the other places where you generate a new name, it's probably safer to have some "fresh name generator" thing (perhaps in the Program) class rather than building names like you do here, to avoid any possible conflicts.

Reply:
Addressed this somehow in next commit ba6fc88.